source/particle/property/elastic_stress.cc

```diff
@@ -0,0 +1,171 @@
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/property/elastic_stress.h>
+#include <aspect/material_model/viscoelastic.h>
+#include <aspect/material_model/visco_plastic.h>
+#include <aspect/initial_composition/interface.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Property
+    {
+      template <int dim>
+      ElasticStress<dim>::ElasticStress ()
+      {}
+
+      template <int dim>
+      void
+      ElasticStress<dim>::initialize ()
+      {
+        AssertThrow((Plugins::plugin_type_matches<const MaterialModel::ViscoPlastic<dim>>(this->get_material_model())
+                     ||
+                     Plugins::plugin_type_matches<const MaterialModel::Viscoelastic<dim>>(this->get_material_model())),
+                    ExcMessage("This particle property only makes sense in combination with the viscoelastic or visco_plastic material model."));
+
+        AssertThrow(this->get_parameters().enable_elasticity == true,
+                    ExcMessage ("This particle property should only be used if 'Enable elasticity' is set to true"));
+
+      }
+
+      template <int dim>
+      void
+      ElasticStress<dim>::initialize_one_particle_property(const Point<dim> &position,
+                                                           std::vector<double> &data) const
+      {
+        // Give each elastic stress field its initial composition if one is prescribed.
+        data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_xx")));
+
+        data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_yy")));
+
+        data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_xy")));
+
+        if (dim == 3)
+          {
+            data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_zz")));
+
+            data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_xz")));
+
+            data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_yz")));
+          }
+
+      }
+
+      template <int dim>
+      void
+      ElasticStress<dim>::update_particle_property(const unsigned int data_position,
+                                                   const Vector<double> &solution,
+                                                   const std::vector<Tensor<1,dim> > &gradients,
+                                                   typename ParticleHandler<dim>::particle_iterator &particle) const
+      {
+        MaterialModel::MaterialModelInputs<dim> in(1, this->n_compositional_fields());
+
```

Comment:
Why not also set `in.cell` now that it is simple to do? Who knows which parts of the material model may depend on the cell?

Reply:
`in.cell = particle->get_surrounding_cell(this->get_triangulation());`

Reply:
@tjhei - @gassmoeller and I had a question regarding obtain a cell iterator for `in.cell`. The line above returns a cell iterator for the triangulation, but what we actually need is a DoF cell iterator for in.cell. Do you know of a way to convert from a triangulation to DoF cell iterator? 

We tried a few permutations (see example below), but couldn't find the right structure. Example:

Reply:
I had another look and the following seems to work for me:
```
        in.current_cell = typename DoFHandler<dim>::active_cell_iterator(*particle->get_surrounding_cell(this->get_triangulation()), 
          &(this->get_dof_handler()));
```
Let me know if that works for you.
I will check if I can add a convenience function to deal.II to avoid this awkward conversion (in future versions).

Reply:
Oh, and you also have to set `in.position` to the particle positions, otherwise the material model will create floating point exceptions (that confused me when testing this).

Reply:
@gassmoeller - this did the trick, thanks! I'll update the PR again once I've finished addressing the remaining comments, tested all the relevant benchmarks, and added a few tests.

Fingers crossed that using the particles to smooth out oscillations in the elastic stresses will improve the VEP non-linear convergence behavior.